Join stuff

RaduBerinde opened the pull request:
@andreimatei

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/raduberinde/cockroach/10)
<!-- Reviewable:end -->

andreimatei commented on docs/RFCS/distributed_sql.md:
```diff
@@ -553,12 +558,175 @@ Processors are generally made up of three components:
+   and distribute copies to all processor instances; the streams for the other
+   table get processed on their respective nodes. This strategy is useful when
+   we are joining a big table with a small table, and can be particularly useful
+   for subqueries, e.g. `SELECT .. WHERE x IN (SELECT ..)`
```
clarify that the table doesn't have an index on `x` (otherwise we'd use route-by-range)

andreimatei commented on docs/RFCS/distributed_sql.md:
```diff
@@ -553,12 +558,175 @@ Processors are generally made up of three components:
+
+The join-by-lookup method involves receiving data from one table and looking up
+corresponding rows from another table. It is typically used for joining an index
+with the table, but they can be used for any join in the right circumstances. We
```
maybe clarify "(i.e. when joining a small `src1` with the PK in `src2`)

andreimatei commented on docs/RFCS/distributed_sql.md:
````diff
@@ -553,12 +558,175 @@ Processors are generally made up of three components:
+```
 
-### Inter-stream ordering
+Note that `JOIN-READER` has the capability of "plumbing" through an input column
````
What's the point of this? That `t2src` doesn't need to actually read `v`? Maybe clarify.

andreimatei commented on docs/RFCS/distributed_sql.md:
```diff
@@ -553,12 +558,175 @@ Processors are generally made up of three components:
+    leaders. We will probably use this strategy for the first implementation.
+
+ 2. it can use routers-by-range to route each input to an instance of
+   `JOIN-READER` on the node for the respective range; the flow of data
```
... on the respective range *of t2*.

andreimatei commented on docs/RFCS/distributed_sql.md:
```diff
@@ -553,12 +558,175 @@ Processors are generally made up of three components:
+    Even with this optimization, it might be too expensive to set up the remote
+    flows if we only end up doing a handful of queries. We can investigate a
+    hybrid approach where we batch some results and depending on how many we
+    have and how many different ranges/nodes they span, we choose
```
Maybe suggest here when the pushing the plan to t2 is particularly useful when the indexes of t1 and t2 are correlated (e.g. date on t1 and autoincrement id on t2)

andreimatei commented:
LGTM

RaduBerinde commented:
Updated.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/distributed_sql.md, line 712 \[r1\]](https://reviewable.io:443/reviews/raduberinde/cockroach/10#-KFFKxbAOaI-FMIbcVJY-r1-712:-KFFrMC-Y0EMcKFJ7ffz:554411405) ([raw file](https://github.com/raduberinde/cockroach/blob/824032a215d52f134a9ae09f4dd055489156011d/docs/RFCS/distributed_sql.md#L712)):*
it's a very generic incomplete example, not worth getting in details about it IMO (we said "can be useful", so it's not necessary the best in all cases like that). I added an ".. AND" to signify there may be other filters (which can "force" a certain index).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/raduberinde/cockroach/10#-:-KFFsFIxA79zWyq6N9S-:-719267080)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[docs/RFCS/distributed_sql.md, line 651 \[r1\]](https://reviewable.io:443/reviews/raduberinde/cockroach/10#-KFFKxbAOaI-FMIbcVJY-r1-651:-KFFtBBukID8NEcrK9q4:-499146009) ([raw file](https://github.com/raduberinde/cockroach/blob/824032a215d52f134a9ae09f4dd055489156011d/docs/RFCS/distributed_sql.md#L651)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/raduberinde/cockroach/10)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/raduberinde/cockroach/10#-:-KFFvGJqnd5HTk7Yf9zy:980077423)*
<!-- Sent from Reviewable.io -->
